### DISCUSSION TITLE
Reverse media queries option

### DISCUSSION THREAD

**Discussion by suiramdev** [tailwindlabs/tailwindcss#7645] [Ideas]
At the moment, as far as I know, we can only make media requests that apply the following class if the screen size is larger than the given size.
I had the idea to add the opposite, by preceding the option with a negative sign: `-md:justify-between`

---

**Solved**
**The feature now exist**
 https://tailwindcss.com/blog/tailwindcss-v3-2#max-width-and-dynamic-breakpoints

**Comment by simonswiss** [comment]
Hey!

You can completely customize how the media queries work in Tailwind CSS:

https://tailwindcss.com/docs/screens

The _default_ is indeed to have "mobile-first", `min-width` breakpoints, but you can quite easily turn those into `max-width` breakpoints if you prefer working this way:

https://tailwindcss.com/docs/screens#max-width-breakpoints

You can't add a `-` at the start to invert the default breakpoints, but you can totally add a custom `-md` max-width breakpoint in your config file to achieve the exact same result!

<img width="640" alt="image" src="https://user-images.githubusercontent.com/485747/155661035-5e021fe6-1359-4956-995b-65c02b794e07.png">

And here's the demo:

https://play.tailwindcss.com/Jrnrq0Ob7I?file=config

Notice that for these breakpoints to override each other correctly, I have the widest breakpoint on top, and the narrowest at the bottom. This will ensure that `-sm` can override `-xl`, due to the order in which the styles will be generated.

**Note**: This is my personal opinion, but I don't suggest mixing `min-` and `max-` width breakpoints when styling a project, as it can get a bit confusing and _conflicting_. If you have a good reason to do it, by all means, go ahead, but I wouldn't start mixing both unless there is a specific reason to do so.

But, effectively, the way I create additional custom breakpoints in this Tailwind Play linked above gives you the exact functionality you where suggesting, where you can prefix your breakpoints with a `-`.

<img width="563" alt="image" src="https://user-images.githubusercontent.com/485747/155662565-9d4ab9f5-4a97-4eca-b7a3-d3ac96f06b0d.png">

Have a read through the entire "Screens" page of the documentation – you can do some pretty cool stuff with these!

**Reply by rootical** [reply]
Spent several hours looking for how to do it! Has almost given up. 
Why the `'key': {max: 'value'}` is so hard to find in the docs and is it even over there? Should be in the Responsive Design Customization section... Anyway glad I've found it. Wasn't smart enough to check the types for config :/

**Reply by the94air** [reply]
I agree. I remember running into the same deal. The documentation around this could be improved.

**Reply by ADTC** [reply]
How do we do this in Tailwind 4 that's using on CSS variables for config now?

```css
  --breakpoint-sm: 40rem;
  --breakpoint-md: 48rem;
  --breakpoint-lg: 64rem;
  --breakpoint-xl: 80rem;
  --breakpoint-2xl: 96rem;
```

These are automatically `width >=` (equivalent to `min-width`).

**Comment by jdockwarder** [comment]
I definitely see the point in this idea. I would support the idea of making this a standard feature in tailwind to decrease the configuration effort. Are there any features interfering with the concept of generate `-BREAKPOINT` variants by default?

**Reply by adamwathan** [reply]
I don't think it needs to be a default personally, it's very straightforward to add it in your own configuration and I don't think most users will need it. I've been building sites with Tailwind for 5 years and haven't found myself wanting a max-width media query at any point for instance.

It would be a bit more complicated than it sounds, because we have to subtract 1px from whatever your min-width media query is, and someone might be using `em` or `rem` for their media query values which makes that math a lot trickier all of a sudden. What is the max-width version of a `56rem` min-width media query for example? 

So I think this is better just handled in your own configuration. We designed the framework to be extremely customizable and flexible for exactly this reason 👍🏻

**Reply by jdockwarder** [reply]
Thanks for the detailed response. It's fun to discuss those topics. I see that there's a lot of complexity coming along with this kind of features.

Regarding your example with rems and ems, wouldn't it be possible to use the calc function? It has pretty good support coverage in browsers: `@media (max-width: calc(56rem - 1px)`.

I'm also not trying to prove a point here, I'm just interested in getting other peoples' opinion on this. I think it's very subjective, whether this makes sense. To me this feels like a very handy implementation or convention over configuration.

Again, I appreciate you take the time for this matter :)

**Reply by suiramdev** [reply]
I think both views are correct. On the one hand, I agree with you, indeed, it could be tricky and we may be in the minority to ask for this change.
I thought it would be cool to have it by default, instead of having to configure it on every project.
It's not a personal configuration, but a choice of project direction that everyone can have, unlike, say, a color code change.

**Reply by suiramdev** [reply]
As @dockwarder said:
You could make any rule with the tailwind configuration file. You wouldn't even need to develop it.
The reason it's good to develop them is that everyone needs them and it would reduce the size of people's configuration files.
So, this feature could.

**Reply by mrleblanc101** [reply]
@adamwathan Why couldn't this be in tailwindcss by default, why JIT mode it doesn't add any weight.
Sometime it's easier to scope you style to mobile only using `-sm:absolute` than to cancel it using `absolute sm:static`
Otherwise, could this be a built-in plugin ? So we don't have to copy paste this config in every tailwindcss project ?

**Reply by the94air** [reply]
To add why this could be useful. I find this utility very helpful when I need to hide a background image on a smaller breakpoint. It's impossible to set this using the standard `min-width` breakpoints and you will have to either use a custom CSS code or add in negative breakpoints. And now that you can add a full URL to the background utility, this is no longer a big issue. It _was_ when we had to inline the URL using style attributes. But yeah, I understand the reason why this shouldn't be added to the default configs and I kind of agree with it. But it could be a nice addition to deal with these cases setting background images. I am also sure there are some special CSS properties that also suffer from the same limitation but I can't remember them at the moment.

**Reply by wongjn** [reply]
Since Tailwind v3.2, [there are `max-*` variants out-of-the-box](https://tailwindcss.com/docs/responsive-design#targeting-a-breakpoint-range) which fulfils this feature suggestion!

**Comment by kocoten1992** [comment]
Thanks for the idea of `-xl`, it a killer feature, It should be in tailwindcss I think.

**Comment by MaKTaiL** [comment]
I would love to have these options included in the official Tailwind as well. It's nice that we can customize it while it isn't official though.

**Reply by mrleblanc101** [reply]
https://github.com/tailwindlabs/tailwindcss/discussions/7645#discussioncomment-4622399

**Reply by MaKTaiL** [reply]
Wow, I hadn't notice this. Thank you for the heads-up!

**Comment by SampsonCrowley** [comment]
I know this is an old thread but is there a good reason not to use: `justify-between md:justify-around` (`hidden md:block`, etc)? setting the value that it should be by default as the "less than md" option, and set what it should change to at "md and above"

**Reply by wongjn** [reply]
For what it's worth, it seems Tailwind advocates the strategy of having the value it should be by default and then setting what it should change to at `md` and above, i.e. [mobile-first](https://tailwindcss.com/docs/responsive-design#working-mobile-first).

**Reply by nicolasrenon** [reply]
Sometimes you can't do that. For example, `align-items`'s default value is `normal`, but Tailwind does not provide a class to set it back to `normal`.

**Comment by alianza** [comment]
I need this for setting `ring-inset` on a button in a small viewport, but there is no utility to enable something like `ring-outset` again. A reverse viewport would fix this!

**Reply by adamwathan** [reply]
We’ve had this feature for almost two years! 🙂

https://tailwindcss.com/blog/tailwindcss-v3-2#max-width-and-dynamic-breakpoints
